### ISSUE TITLE
Trying pre-processing my dataset

### ISSUE THREAD

**Issue by Itzikwa** [frankkramer-lab/MIScnn#15]
Hey Dr. Mueller,

I had install your Python package on my computer, and it's seems to be working pretty good (I checkd it out by using the KiTS19 example). now the next step is to use MIScnn on my datasets.

In our research we are triyng to segment and clasify pathology (cancer, cyct etc.) in the jaw. For this we are using CT scans which are in PNG format (the original images) and JSON fies (represent the cordinates of the desirable segmentation). 

is There a chance to use your implementation so it will be suitable to our needs?

best regards, Itzik

**Comment by muellerdo** [comment]
Heyho Itzik,

glad to hear that you are interested in using MIScnn for your research.

As @MichaelLempart correctly answered, the Image IO Interface is a good starting point for your requirements, because your imaging data is probably 2D if in PNG format.

> @Itzikwa said:
is There a chance to use your implementation so it will be suitable to our needs?

I'm quite optimistic that we get a medical image segmentation running for your data set! :)

> @Itzikwa said:
> In our research we are triyng to segment and clasify pathology (cancer, cyct etc.) in the jaw. For this we are using CT scans which are in PNG format (the original images) and JSON fies (represent the cordinates of the desirable segmentation).

The storage of segmentation in JSON is quite uncommon. The provided MIScnn Image IO interface for data loading expects the segmentation in a PNG file format in the following file structure:
```
data/
     id001/imaging.png
              segmentation.png
    id002/imaging.png
              segmentation.png
     id003/imaging.png
              segmentation.png
     ...
```
We have two options, now.
1. Preprocessing apart from MIScnn into the shown file structure and using the Image IO interface, afterwards.
2. The implementation of a custom IO interface for data loading.

I would recommend implementing a custom IO interface. 
The working steps would be as follows:
- Copy&paste the Image IO interface as code base
- Optional: Modify the initialize() and load_image() function that it suits your file structure.
- Modify the load_segmentation() function in order to load & parse a JSON file. At the end, we need to return a NumPy matrix with the same shape as the image including the classification for each pixel/voxel according to the JSON file.

I hope that I was able to help you.
If you need further help in building this custom IO interface,
please don't hesitate to ask.

Cheers,
Dominik

**Comment by Itzikwa** [comment]
Thank you for the quick answer! I also checked out the Cell Tracking example and it seems to be very usefull...

But there is one mainly point that i didn't get at all. My data contains like 30 cases of CT scans, and every scan was seperated to many 2D pics. If i had understand correctly, the image IO interface designed for processing independent set of 2D pics. aren't many utilities can be achieved by using a 3D model in this case?

I hope i was clearly enough...

**Comment by muellerdo** [comment]
Ah, sure you can combine 2D image planes back to a 3D volume. 
MIScnn does not provide a data loader for this scenario, but it could be implemented in a custom IO interface. 

But you have to be aware that you will introduce a bias into your data which can lead to worse performance than 2D architectures.
Normally, you have further information in 3D medical imaging data in NIfTI or DICOM format. Most important is the slice thickness / voxel spacing. 

Here are some good explanations of slice thickness:
http://www.mrishark.com/slice-thickness.html
https://www.materialise.com/en/faq/what-difference-between-slice-thickness-and-slice-increment

Why is this important for us?
The voxel spacing have a strong impact on how well a model can fit and its final performance.
Because CT scans can have difference voxel spacings, you perform voxel space normalization (resampling) as preprocessing. 
The big problem now is that without this information, we are not able to do this. 

You can, of course, ignore this bias, combine the 2D images to a 3D volume and artificially define the voxel spacing as 1x1x1. 

Sadly, I don't have any paper in my mind right now, which compared the performance between 2D images and 3D without resampling. I personally would guess that the 2D performance is better.
But it would be quite interesting to test this out!

**Long story short:** 
It would be the best if you have the original DICOM or NIfTI files which includes the voxel spacing.
Otherwise, I personally would recommend to try out a 2D approach. But, like I said, I don't have a paper in mind right now, which compared the performance of these two approaches, so it is absolute legit if you want to try this out. :)

> I hope i was clearly enough...

Don't worry, it was explained quite well. :)

**Comment by Itzikwa** [comment]
Thanks again! Actually i do have the original dicom files, but i thought it will be much easier to segment manually 2D images...

So, if i understand you correctly, your suggestion is to convert my dicom files to nifti format, and then to use the nifti io interface?

**Comment by muellerdo** [comment]
If you have the original files, I would recommend using the full 3D information. :)

Either by converting the DICOM files into NIfTI and then using the NIfTI IO Interface, as you correctly already suggested.

Or implementing a custom IO interface for DICOM data (Pydicom is a nice package for processing DICOM files).

If you don't want to worry about writing your own IO interface, I would recommend converting DICOM to NIfTI.

**Comment by muellerdo** [comment]
I will close this issue due to inactivity.

If this issue was not solved, yet, don't hesitate to open it up, again.
